node/src/network.rs

```diff
@@ -293,88 +417,304 @@ impl NetworkTaskChannel {
+            MpcMessage {
+                task_id: self.task_id,
+                kind: MpcMessageKind::Computation(data),
             },
         )
     }
 
-    /// Receives a message from another participant in the MPC task.
-    ///
-    /// Messages from the same sender are guaranteed to be received in the same order they were
-    /// sent. However, messages from different senders may be received in arbitrary order.
-    ///
-    /// Returns an error if the networking client is dropped (during node shutdown).
+    pub fn is_leader(&self) -> bool {
+        self.my_participant_id == self.leader
+    }
+
+    /// Waits for all participants involved in this task to be bidirectionally connected to us.
+    /// This should be called at the beginning of the computation if:
+    ///  - This is a leader-centric computation, and we are a follower.
+    ///    (Rationale: the leader already determined that the participants are online. We wait
+    ///     for connections because even though the leader has connected to everyone, it is
+    ///     possible we have yet to establish connections to everyone. We assume that if the
+    ///     leader can connect to two nodes B and C, then B and C can also connect to each other.)
     ///
-    /// This future may never resolve if the MPC computation fails to progress (i.e. all clients
-    /// decide they need to receive a message before sending one). It is up to the caller to
-    /// implement a timeout mechanism. However, if we notice that not all participants of the
-    /// computation are online anymore, this method will return an error soon.
-    pub async fn receive(&mut self) -> anyhow::Result<MpcPeerMessage> {
-        loop {
+    /// This should NOT be called if:
+    ///  - We are the leader. This is redundant because the leader already queried the alive
+    ///    participants when making the choice of participants to use in the computation.
+    async fn wait_for_all_participants_connected(&self) -> anyhow::Result<()> {
+        for &participant in &self.participants {
+            if participant == self.my_participant_id {
+                continue;
+            }
+            tracking::set_progress(&format!("Waiting for connection to {}", participant));
+            self.transport_sender
+                .connectivity(participant)
+                .wait_for_connection(self.connection_versions[&participant])
+                .await?;
+        }
+        tracking::set_progress("All participants connected");
+        Ok(())
+    }
+
+    /// This is called at the end of a leader-centric computation to communicate the result to
+    /// other parties. For leaders, only failures are communicated to followers; for followers,
+    /// failures are communicated to the leader, and if leader waits for success, also
+    /// communicate successes to the leader.
+    fn communicate_result<T>(
+        &self,
+        result: &anyhow::Result<T>,
+        leader_waits_for_success: bool,
```

Comment:
Under what scenarios does a leader wait for success?

Reply:
In my new code I documented it.